Detour/Source/DetourNavMesh.cpp

```diff
@@ -616,12 +616,48 @@ void dtNavMesh::baseOffMeshLinks(dtMeshTile* tile)
 	}
 }
 
+void dtNavMesh::closestPointOnDetailEdge(const dtMeshTile* tile, const dtPoly* poly, const float* pos, float* closest) const
+{
+	const unsigned int ip = (unsigned int)(poly - tile->polys);
+	const dtPolyDetail* pd = &tile->detailMeshes[ip];
+
+	float result[3];
+	dtVcopy(result, pos);
+
+	float dmin = FLT_MAX;
+	for (int i = 0; i < pd->triCount; i++)
+	{
+		const unsigned char* tris = &tile->detailTris[(pd->triBase + i) * 4];
```

Comment:
tris[3] is flag indicating if the edge is part of the nav poly contour, see:
https://github.com/recastnavigation/recastnavigation/blob/master/DebugUtils/Source/DetourDebugDraw.cpp#L104

With that information, it should be possible to iterate over edges only once, by checking that either the contour flag is set, or indices of a segment are in ascending order. Or iterate only the boundary edges.

Reply:
Note: The border flag is not 100% accurate, there can be false positives, which should be ok in this case.

Reply:
Gave it a go. I split them into two functions `closestPointOnDetailedContour` and `closestPointOnDetailEdge`. The latter could be used for both cases, but the common case (where point is outside poly) needs the former, and it can be written more efficiently. I will do some benchmarks and see whether it makes a difference. How does it look?

Reply:
May I suggest `closestPointOnDetailEdge` -> `closestPointOnDetailEdges`, as it does iterate over all edges, not just one. Maybe `closestPointOnDetailedContour` -> `closestPointOnDetailContour` for consistency.

Reply:
I have added just a single closestPointOnDetailEdges taking a template parameter that controls the behavior.

In terms of speed, iterating only boundary edges makes a massive difference (around 100% faster to iterate only boundary edges vs all edges).

In comparison to the old, results are a bit mixed. I tested with 10 million closestPointOnPoly queries on a large poly (11 detail triangles) and small poly (1 triangle).
When the point is inside, the new version is about 20% faster for both polys, I guess because we no longer calculate closest "non-detailed" edge.

When the point is outside, the old version is quite a lot faster on the big poly, depending on which edge is closest, with the old version takes between 55% and 90% of the time of the new version.
For the small poly the new version is 0-20% faster depending again depending on the edge.

Anyway, since this is a bug fix I guess the trade off is worth it. I also think it is likely that this function is called with points inside the poly more often than outside.

Reply:
I think those numbers makes sense. Did you check how often the `dtClosestHeightPointTriangle` failed and the edge fall back was used for the inside case?

Reply:
I did not try "provoking" this situation and never saw it happen, but I was clicking manually and that makes it really unlikely to happen. I can try with some edge/junction points tonight and see if they trigger the case.

Reply:
I used to have a test level which had big polygons on an undulating terrain quite far from origo. It might had had some big sliver polygons too. I think that level had problems inside the polygon. I tested it using a dozen agents.

Reply:
I tried the following situation and moving the target around a bit, but still didn't trigger the case:
![image](https://user-images.githubusercontent.com/7887810/54072274-dc3ced80-4278-11e9-8bc0-5f8d277b9add.png)

This is with origin at 5000, 0, 5000.

Reply:
You could try bigger tile size (i.e. 4 times bigger) and add few obstacles in there so that you're more likely to get sliver nav polys.

Reply:
I was able to hit the case after adding some obstacles. However, it is not hitting it internally in polygons, in all the cases it was on the border of a polygon (i.e. `dtPointInPolygon` disagrees with the triangle check in `dtClosestHeightPointTriangle`). That is, I could never get a breakpoint inside the if here to be hit:
```cpp
float closest[3];
closestPointOnDetailEdges<false>(tile, poly, pos, closest);
*height = closest[1];

float closest2[3];
closestPointOnDetailEdges<true>(tile, poly, pos, closest2);
if (dtVdistSqr(closest, closest2) > 0.0001 * 0.0001)
{
	int ii = 0; // never hit
}
```
I think I will leave it as is since this should be the most robust way to do it and it almost never happens anyway. I have also added the mesh I have made to test to this PR.